mos-platform/common/include/fixed_point.h

```diff
@@ -0,0 +1,301 @@
+/// point number. The first number represents the number of bits used to hold the
+/// Integer portion of the value, and the second 8 bits hold the fractional portion.
+/// This has some very useful properties for the 6502 such as when adding two fixed
+/// point numbers with the same size, the math is the simply 16 bit math. Another
+/// benefit is should you only want to use the integer portion of the value, accessing
+/// this comes with no additional cost because its merely the high bytes. This
+/// property is not true for all combination of sizes, but in general, it still is
+/// reasonably quick to access the higher byte by shifting as needed.
+///
+/// In a little bit more detail, a fixed point number can be considered a regular
+/// 16 bit number that is implicitly divided by the size of the fractional portion.
+/// With our 8.8 example, the Fractional portion can represent 256 values total,
+/// meaning the value in the fractional part represents a number divided by 256.
+/// As a practical example, a 16 bit number `0x5060` , and when divided by `0x100`
+/// yields `0x50` with a remainder of `0x60`. This class makes that divider a
+/// part of the type of the class, allowing seamless math operations as if its a
+/// normal integer, but with the correct shifting when combining with other values.
+template<uint8_t ISize, uint8_t FSize, bool Signed = true> class Value;
+
+template<uint8_t ISize, uint8_t FSize, bool Signed>
+class Value {
+protected:
+
+  // Used to calculate the bitmasks for both the integer part and fractional part
+  template <typename R>
+  static constexpr R BitMask(unsigned int const onecount) {
+    return static_cast<R>(-(onecount != 0))
+      & (static_cast<R>(-1) >> ((sizeof(R) * CHAR_BIT) - onecount));
+  }
+
+  // Size of the underlying storage for the different views of the data.
+  // The total combined storage will be used for representing the actual value.
+  using IntType = std::conditional_t<Signed, 
+                                     signed _BitInt(((ISize + 7) / 8) * 8),
+                                     unsigned _BitInt(((ISize + 7) / 8) * 8)>;
+  using FracType = std::conditional_t<Signed,
+                                      signed _BitInt(((FSize + 7) / 8) * 8),
+                                      unsigned _BitInt(((FSize + 7) / 8) * 8)>;
+  using StorageType = std::conditional_t<Signed,
+                                         signed _BitInt((((ISize + FSize) + 7) / 8) * 8),
+                                         unsigned _BitInt((((ISize + FSize) + 7) / 8) * 8)>;
+  StorageType val;
+  static constexpr unsigned long INT_MASK = BitMask<unsigned long>(ISize) << FSize;
+  static constexpr unsigned long FRAC_MASK = BitMask<unsigned long>(FSize);
+
+public:
+
+  // Constructors
+
+  /// Common constructor for creating a fixed point value from an integer value
+  [[clang::always_inline]] constexpr explicit Value(IntType i) {
```

Comment:
The general zeitgeist in C++ is to have lossless, cheap conversions be implicit. This appears to be the case for `IntType` -> `FixedPoint`. Having this implicit would allow you to e.g. pass an integer to a function that expects a fixed point; compare passing an integer to a function that expects a float.